Add Tracy as optional dependency for profiling

topolarity opened the pull request:
This PR adds libTracyClient as an optional build dependency, to be used for profiling the base Julia runtime. Support can be enabled by adding `WITH_TRACY := 1` to Make.user

Here's a small example:
```c
TracyCZoneN(ctx, "jl_subtype_env", true);
int r = jl_subtype_env(x, y, NULL, 0);
TracyCZoneEnd(ctx);
return r;
```

In a follow-up PR, I plan to update the `JL_TIMING` macros to hook into Tracy when available and perhaps add some additional instrumentation around the loading system.

topolarity committed:
Add Tracy as optional dependency

This commit adds libTracyClient as an optional build dependency,
to be used for profiling the base Julia runtime. Support can be
enabled by adding `WITH_TRACY := 1` to Make.user

Here's an example instrumentation:
```c
    TracyCZoneN(ctx, "jl_subtype_env", true);
    int r = jl_subtype_env(x, y, NULL, 0);
    TracyCZoneEnd(ctx);
    return r;
```

This only adds Tracy itself. In a follow-up commit, I'll update the
JL_TIMING macros to hook into Tracy when available.

topolarity committed:
Update Tracy JLL to v0.9.0+1

This version includes the full set of feature flags that we use to
customize Tracy to, e.g., accept on-demand connections instead of
queuing profiling data until receiving a connection.

topolarity committed:
Setup `pthread_sigmask` before loading libjulia-internal

Loading libjulia-internal and libjulia-codegen can trigger constructors
in their dependencies, including some which may spawn threads (e.g.
Tracy). It is important that these not be configured to receive any
signals that libjulia-internal is expecting to `sigwait()` on, esp. when
their default action is to terminate the process.

Technically, the masked signal set should probably be synchronized with
the `sigwait_sigs` set in `signals-unix.c`, but for now this simply masks
all signals.

topolarity commented on cli/loader_lib.c:
```diff
@@ -353,6 +353,17 @@ static char *libstdcxxprobe(void)
 void *libjulia_internal = NULL;
 void *libjulia_codegen = NULL;
 __attribute__((constructor)) void jl_load_libjulia_internal(void) {
+#if defined(_OS_LINUX_)
```
@staticfloat I've added you as a reviewer because of this change to the cli loader.

Let me know if you'd like this done in a different way :+1:

staticfloat replied:
To be explicit, I think setting the signals to `SIG_BLOCK` without having a `sigwait()` handler running yet is okay because we only have two choices at this point:

* Execution continues to the point that we start up our signal handler, at which point any queued signals will be processed.
* Execution halts before then due to a fatal error, at which point signal handling is moot.

topolarity replied:
I thought of one defect: This was leaving more signals blocked than necessary, including signals that we don't sigwait() on which Julia code or an application embedding Julia may wish to use. I've pushed a fix to restore the signal mask at the end of this constructor, which should fix that problem.

P.S. This also made me realize that for Julia-in-Julia, our signal handling is probably totally non-deterministic about which Julia runtime receives a signal.

staticfloat replied:
> This also made me realize that for Julia-in-Julia, our signal handling is probably totally non-deterministic about which Julia runtime receives a signal.

We could define an environment variable that disables signal handling in child Julia's.

topolarity replied:
How about jl_init_options instead? I can spin up a PR for that on Monday

staticfloat commented on deps/libtracyclient.mk:
```diff
@@ -0,0 +1,72 @@
+LIBTRACYCLIENT_CMAKE += -DTRACY_NO_CRASH_HANDLER=ON
+LIBTRACYCLIENT_CMAKE += -DTRACY_ON_DEMAND=ON
+
+$(LIBTRACYCLIENT_SRCCACHE)/cmake-patch-applied: $(LIBTRACYCLIENT_SRCCACHE)/source-extracted
+ifneq ($(OS),WINNT)
+	echo "target_compile_definitions(TracyClient PUBLIC __STDC_FORMAT_MACROS)" >> $(LIBTRACYCLIENT_SRCCACHE)/CMakeLists.txt
+else
+	echo "target_compile_definitions(TracyClient PUBLIC WINVER=0x0602 _WIN32_WINNT=0x0602)" >> $(LIBTRACYCLIENT_SRCCACHE)/CMakeLists.txt
+endif
+	echo 1 > $@
+
+$(LIBTRACYCLIENT_SRCCACHE)/libTracyClient-freebsd-elfw.patch-applied: $(LIBTRACYCLIENT_SRCCACHE)/cmake-patch-applied
+	cd $(LIBTRACYCLIENT_SRCCACHE) && \
+		patch -p1 -f < $(SRCDIR)/patches/libTracyClient-freebsd-elfw.patch
+	echo 1 > $@
+
+$(LIBTRACYCLIENT_SRCCACHE)/libTracyClient-no-crash-handler.patch-applied: $(LIBTRACYCLIENT_SRCCACHE)/libTracyClient-freebsd-elfw.patch-applied
+	cd $(LIBTRACYCLIENT_SRCCACHE) && \
+		patch -p1 -f < $(SRCDIR)/patches/libTracyClient-no-crash-handler.patch
+	echo 1 > $@
```
@vtjnash this is a good example of a case where I think it makes more sense to unpack source within `$(BUILDDIR)`, as opposed to `$(SRCCACHE)`, since we're patching things.  Do you agree?

topolarity replied:
Sounds good to me - I only did it this way because that's how we handle patches for LLVM right now.

I assume you effectively would like to see https://github.com/JuliaLang/julia/pull/48830 done for this Makefile?

staticfloat replied:
Yes, correct

inkydragon replied:
xref: https://github.com/JuliaLang/julia/issues/31530

topolarity committed:
Use BUILDDIR instead of SRCCACHE for libTracyCore

This is the right strategy according to #31530, so that the patching
process doesn't hit any races between multiple builds running from the
same SRCCACHE.

KristofferC commented:
Would be good to have a small section about this in the devdocs at some point, as a reference.
